<style>
@media print
{
h2 {page-break-before:always}
}
</style>

# Homework 03

### Brown University  
### DATA 1010  
### Fall 2019

## Problem 1

Eytan applied gradient descent to the function $f(x) = (x-3)(x-1)(x+2)(x+3)$ starting from $x_0 = -4$, and he found that his algorithm failed to converge because $x_2 = -4$, $x_4 = -4$, and so on. In other words, the iterates alternated between $-4$ and another value. What learning rate did he use?

Hint: set up an equation for $\epsilon$ and use SymPy to solve it.

In [35]:
using SymPy

In [36]:
@vars x
@vars e
f(x) = (x-3)*(x-1)*(x+2)*(x+3)
df(x) = 4x^3 + 3x^2 - 22x - 9
h(e) = e*df(-4) + e*df(-4-e*df(-4))
float.(SymPy.solve(h(e),e))[4]

0.05849294319646785913848585943736725208548405802872302341179598184539510480279844

There are 4 values and we choose the positive real number. 

## Problem 2

Consider the following random experiment: you flip a coin, and if it turns up heads, you roll a die. If it turns up tails, then you draw a ball from an urn containing balls labeled 1 to 11.

Define a sample space $\Omega$ for this random experiment.

**Solution:**    
$$\{(H, 1), (H, 2), \dots, (H, 6), (T, 1), (T, 2), \dots, (T, 11)\}$$

## Problem 3

Suppose that $E$, $F$, and $G$ are events. Come up with events $\tilde{E}$, $\tilde{F}$, 
and $\tilde{G}$ which are pairwise disjoint and which satisfy... 

<center>$ E = \tilde{E}, E\cup F = \tilde{E}\cup\tilde{F}$, and $E \cup F \cup G = \tilde{E}\cup\tilde{F}\cup\tilde{G}.$ </center>

<img src="IMG_0092.png" style="height: 250px;">

Let $E = \tilde{E} = (1,2,3,4)$,    
$F = (3,4,5,6), \tilde{F} = (5,6)$,     
$G = (2,3,6,7), \tilde{G} = (7)$

## Problem 4

Suppose that $A$ is the event that the high temperature in Providence next Tuesday is at least $65$ degrees Fahrenheit, let $B$ be the event that the high temperature in Providence next Tuesday is at least $60$ degrees Fahrenheit, and let $C$ be the event that it rains in Providence next Tuesday. Write each of the following events using the sets $A$, $B$, and $C$ and the operations $\cap$, $\cup$, and $ ^\mathrm{c}$.

(a) It will be less than $60$ degrees all day and rainy next Tuesday in Providence.

(b) The high temperature in Providence next Tuesday will be at least $60$ degrees but not as high as $65$ degrees.

(c) In Providence next Tuesday,it will either be dry or warm (where warm is defined to mean "daily high of at least $65$ degrees").

(d) The daily high temperature in Providence next Tuesday will be higher than $65$ degrees and less than $60$ degrees.

(a) $B^\mathrm{c} \cap C$.    
(b) $B \cap A^\mathrm{c}$.    
(c) $C^\mathrm{c} \cup A$.    
(d) $A \cap B^\mathrm{c} = \phi$.

## Problem 5

Write a Julia program to find the exact value of the probability that in eight flips of a fair coin, no two consecutive flips turn up heads.

Two hints: (i) you can write a recursive function to generate all possible outcomes, and (ii) if you store your flip sequences as strings, you can use the function `occursin` to check for the substring `HH` .

In [43]:
# 0 = 'H' & 1 = 'T'
s = []
for f1 in 0:1
    for f2 in 0:1
        for f3 in 0:1
            for f4 in 0:1
                for f5 in 0:1
                    for f6 in 0:1
                        for f7 in 0:1
                            for f8 in 0:1
                                e = string(f1) * string(f2) * string(f3) * string(f4) * string(f5) * string(f6) * string(f7) * string(f8) 
                                push!(s,e)
                            end
                        end
                    end
                end
            end
        end
    end
end
count = 0
for set in s
    if !occursin("00", set)
        count = count + 1
    end
end
count/length(s)

0.21484375

## Problem 6

The 52 cards in a standard deck (13 spades, 13 clubs, 13 diamonds, 13 hearts), are dealt into four hands of 13 cards each. What is the probability that one of the hands all of the hearts?

**Solution:**    
$ 4\cdot \frac{13}{52} \cdot \frac{12}{51} \cdot \frac{11}{50} \cdot \dots \cdot \frac{1}{40} = 6.2990781e-12$

## Problem 7

Consider a random independent sequence of letters uniformly distributed in {a, b, . . . , z}. Use Monte Carlo simulation to estimate the expected number of letters that appear in the sequence up to the first appearance of aa.
Repeat with ab in place of aa. Based on your findings, is the expected time to the first aa different from the expected time to the first ab?
 

In [ ]:
# chr = 'a' : 'z'
# count = []
# time = 100
# for i in 1:time
#     s = ""
#     while !occursin("aa", s)
#         i = rand(1:26)
#         s = s * chr[i]
#     end
#     push!(count, length(s)-2)
# end

# mean(count)

In [40]:
chr = 'a' : 'z'
count = []
time = 100000
for i in 1:time
    s = "  "
    while s[length(s)-1:end] != "aa"
        i = rand(1:26)
        s = s * chr[i]
    end
    push!(count, length(s)-4)
end

mean(count)

705.11574

The expected number of letters that appear in the sequence up to the first appearance of aa is 705

In [41]:
count2 = []
time = 100000
for i in 1:time
    s = "  "
    while s[length(s)-1:end] != "ab"
        i = rand(1:26)
        s = s * chr[i]
    end
    push!(count2, length(s)-4)
end

mean(count2)

671.00168

The expected number of letters that appear in the sequence up to the first appearance of aa is 671

## Problem 8

Suppose that many independent samples are drawn from the joint distribution of two random variables $X$ and $Y$, and the results are as shown in the first figure below. Sketch, by shading, your best guess of the density function of the distribution of $(X,Y)$. Also sketch a graph of the function $x\mapsto \mathbb{E}[Y | X = x]$.

<center>
    <div class="row">
    <img src="images/scatter1.svg" style="height: 250px;">
    <img src="images/scatter2.svg" style="height: 250px;">
    <img src="images/scatter3.svg" style="height: 250px;">
    <img src="images/scatter4.svg" style="height: 250px;">
    </div>
</center>

Repeat the exercise for each of the remaining figures.

<img src="IMG_0088.png" style="height: 250px;">

<img src="IMG_0089.png" style="height: 250px;">

<img src="IMG_0090.png" style="height: 250px;">

<img src="IMG_0091.png" style="height: 250px;">